packages/pds/src/api/com/atproto/repo/importRepo.ts

```diff
@@ -0,0 +1,128 @@
+import { Server } from '../../../../lexicon'
+import AppContext from '../../../../context'
+import { ActorStoreTransactor } from '../../../../actor-store'
+import { TID } from '@atproto/common'
+import {
+  Repo,
+  WriteOpAction,
+  getAndParseRecord,
+  readCarStream,
+  verifyDiff,
+} from '@atproto/repo'
+import { InvalidRequestError } from '@atproto/xrpc-server'
+import { CID } from 'multiformats/cid'
+import PQueue from 'p-queue'
+import { AtUri } from '@atproto/syntax'
+import { BlobRef, LexValue, RepoRecord } from '@atproto/lexicon'
+
+export default function (server: Server, ctx: AppContext) {
+  server.com.atproto.repo.importRepo({
+    auth: ctx.authVerifier.accessNotAppPassword,
+    handler: async ({ input, auth }) => {
+      const did = auth.credentials.did
+      if (!ctx.cfg.service.acceptingImports) {
+        throw new InvalidRequestError('Service is not accepting repo imports')
+      }
+      await ctx.actorStore.transact(did, (store) =>
+        importRepo(store, input.body),
+      )
+    },
+  })
+}
+
+const importRepo = async (
+  actorStore: ActorStoreTransactor,
+  incomingCar: AsyncIterable<Uint8Array>,
+) => {
+  const now = new Date().toISOString()
+  const rev = TID.nextStr()
+  const did = actorStore.repo.did
+
+  const { roots, blocks } = await readCarStream(incomingCar)
+  if (roots.length !== 1) {
+    throw new InvalidRequestError('expected one root')
+  }
+  const currRoot = await actorStore.db.db
+    .selectFrom('repo_root')
+    .selectAll()
+    .executeTakeFirst()
+  const currRepo = currRoot
+    ? await Repo.load(actorStore.repo.storage, CID.parse(currRoot.cid))
+    : null
+  const diff = await verifyDiff(
+    currRepo,
+    blocks,
+    roots[0],
+    undefined,
+    undefined,
+    { ensureLeaves: false },
+  )
+  diff.commit.rev = rev
+  await actorStore.repo.storage.applyCommit(diff.commit, currRepo === null)
+  const recordQueue = new PQueue({ concurrency: 50 })
+  for (const write of diff.writes) {
+    recordQueue.add(async () => {
+      const uri = AtUri.make(did, write.collection, write.rkey)
+      if (write.action === WriteOpAction.Delete) {
+        await actorStore.record.deleteRecord(uri)
+      } else {
+        let parsedRecord: RepoRecord
+        try {
+          const parsed = await getAndParseRecord(blocks, write.cid)
+          parsedRecord = parsed.record
+        } catch {
+          throw new InvalidRequestError(
+            `Could not parse record at '${write.collection}/${write.rkey}'`,
+          )
```

Comment:
I think any errors in here will end-up unhandled and cause a crash.  What we probably want to do is setup an `AbortController` and user `p-queue``'s support for signals to thread it through and bail on work once there's a failure.  One the `diff.writes` loop ends and the queue is determined idle we can call `signal.throwIfAborted()` to raise the original error if one occurred.

Reply:
oof whatdya know 🙃 

i'll fix this up

Reply:
probably worth a quick peek: https://github.com/bluesky-social/atproto/pull/2179/commits/153c90b145072db1a8853ff87ace414ce4f61adf

Reply:
Looks good 👍

Reply:
yup i gave it a test & it worked how you'd expect 👍